_posts/2025-11-06-sealing.markdown

```diff
@@ -0,0 +1,202 @@
+
+The first use of these most programmers will make is to allocate memory.
+If you call `malloc`, this is a thin compatibility wrapper around `heap_allocate`, which requires an allocator (software) capability as an argument.
+The allocator capability is a sealed object that contains a quota.
+It authorises you to allocate memory until your quota is exhausted.
+It also authorises you to free objects that were allocated from your quota and reclaim the quota.
+Similarly, when you create a connected TCP socket, you pass a capability that authorises you to connect to a specific host and port.
+
+These static capabilities can be inspected with [CHERIoT Audit](/rtos/firmware/auditing/2024/03/01/cheriot-audit.html) so you can write policies that say things like 'these four compartments, between them, can't allocate more than 16 KiB of RAM' or 'this compartment may connect to my cloud back end, but nowhere else'.
+
+# Sealed handles give trivial flow isolation
+
+Sealed capabilities are passed out and back as opaque values and can be unsealed without any global state beyond the sealing key.
+A lot of compartments that use them have no global mutable state.
+If you pass a sealed capability to a TLS session into the TLS compartment, that compartment can unseal it and then see the state of your TLS session.
+
+While operating on your TLS session, that compartment does not hold any capabilities to other TLS sessions.
+If an attacker gains arbitrary code execution in the TLS compartment while operating on one TLS flow, that doesn't give them the ability to attack another one.
+
+The same applies to much simpler compartments.
+The message-queue compartment, which provides secure message queues and streams between compartments, has the same property.
+The only state that it operates over is reached via the sealed queue pointer or one of the arguments.
+If you compromise this compartment, you can tamper with a queue that you hold an endpoint handle for, but not any other queue.
+
+# Sentries build on sealing
+
+CHERI also has a notion of a *sealed entry* (sentry) capability.
+These are sealed capabilities with a special otype that allows them to be used as jump targets.
+When you jump to a sentry, it is implicitly unsealed.
+
+Morello had several variants of sentries, including some designed for descriptors where the sentry was actually a pointer to a pair of a code and data capability.
+The jump would load the data capability and branch to the code capability.
+This underlying mechanism is very flexible and there is a lot more research to be done on how it can be used in the future on big systems.
+
+In CHERIoT RTOS, all library functions, including the switcher (which handles cross-compartment calls) are provided as sentries.
+The use of a sentry otype is an *attestation* from the loader that this really is a valid jump target, which code can check.
```

Comment:
This is only sort of true, in ways that don't make sense to discuss until after the "rich sentries", so maybe move this sentence down?